### PR TITLE
Make `Path` carry the expected result type to allow fully typed generic components

### PR DESCRIPTION
**Pull request by julienfouilhe**
I'm in no way a Typescript expert but I wanted to be able to create generic form components.
The way I thought it could work was if the `Path` type returns all paths for a compatible type instead of all possible paths.

So you can write components like this (where expected type would actually be infered from a yup/zod/iots... schema)

```typescript
type ExpectedType = { fileName: string; url: string };

const GenericFileUpload = <FormValues extends FieldValues>({
  name,
  control,
}: {
  name: FieldPath<FormValues, ExpectedType>;
  control: Control<FormValues>;
}) => {
  useController({
    name,
    control,
    defaultValue: { fileName: 'My test file', url: 'https://localhost/image.jpg' }, // this is typed and will fail if you pass it something that does not extend ExpectedType
  });

  return null;
};
```

Used like this:

```typescript
interface FormValues {
  invalid: { 
    subpath: string;
  };
  file: {
    fileName: string;
    url: string;
  };
}

const { control } = useForm<FormValues>();

// this will pass
<GenericFileUpload name="file" control={control} />
// while this will not, as expected
<GenericFileUpload name="invalid" control={control} />
```

⚠️ This does not handle all use cases yet and I may have went a little overboard where I put it but I wanted to check if you thought this could be a good idea before continuing?

### CODE REVIEW

**Comment by bluebill1049** [comment]
Thanks for the contribution! I will take a look at it. cc(@kotarella1110)

**Comment by bluebill1049** [comment]
looks like this break the build. 

Screen Shot 2021-09-06 at 8 17 39 pm

**Comment by julienfouilhe** [comment]
@bluebill1049 Yes I know, this PR is not finished, I just wanted to know if you thought this was a good idea or if you had some context that would make this a no go.

**Comment by bluebill1049** [comment]
I like this idea, I think the reusable component is one of the weaknesses of the current strict type with template literal. @kotarella1110 let me know your thoughts.

related issue: https://github.com/react-hook-form/react-hook-form/issues/5055

File: src/controller.tsx [hunk 1]
```diff
@@ -3,9 +3,13 @@ import { useController } from './useController';
 
 const Controller = <
   TFieldValues extends FieldValues = FieldValues,
-  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
+  TResult = unknown,
```
Comment (bluebill1049) [inline src/controller.tsx:6]: `TFieldValue` sounds better?

File: src/__tests__/useController.test.tsx [hunk 2]
```diff
@@ -26,6 +26,37 @@ describe('useController', () => {
...
   });
 
+  it('should render generic component correctly', () => {
+    type ExpectedType = { test: string };
+
+    const Generic = <FormValues extends FieldValues>({
+      name,
+      control,
+    }: {
+      name: FieldPath<FormValues, ExpectedType>;
```
Comment (bluebill1049) [inline src/__tests__/useController.test.tsx:36]: I guess the only cons/pros is the user will be able to overwrite/skip the type check.

**Comment by slbls** [comment]
@julienfouilhe, is there anything I can do to help with this PR? The ability to type generic components would be a great addition to React Hook Form and directly useful in projects of my own.

**Comment by bluebill1049** [comment]
> @julienfouilhe, is there anything I can do to help with this PR? The ability to type generic components would be a great addition to React Hook Form and directly useful in projects of my own.

Thanks 👍 will let u know if we need some help. Give us sometimes on this one.

**Comment by bluebill1049** [comment]
hey @barrymay may i get your thoughts on this PR as well?

File: app/src/setError.tsx [hunk 3]
```diff
@@ -45,7 +45,7 @@ const SetError: React.FC = () => {
         message: 'Minlength is 10',
       },
     ].forEach(({ name, type, message }) =>
-      setError(name as FieldPath<FormInputs>, { type, message }),
+      setError(name as FieldPath<FormInputs, any>, { type, message }),
```
Comment (barrymay) [inline app/src/setError.tsx:48]: can we use unknown instead of any here?

**Comment by barrymay** [comment]
@julienfouilhe Agreed this would be a great add to the library, although you really want to make sure it has good defaults for everyone, while doing what you say it's gonna do.  It's certainly something I've been interested in!

I'll have to play with this branch a bit before I comment more ;)

**Comment by kotarella1110** [comment]
I like this suggestion.
If it is difficult to fix, I can help you 👍🏻

**Comment by bluebill1049** [comment]
I think this is one of the missing pieces right now for reuseable components, would be great to push this one moving forward since both @barrymay and @kotarella1110 agrees too.

**Comment by julienfouilhe** [comment]
Closing in favor of #6562

**Pull request closed without merging**